AX: AXPropertyName::Language can become out-of-date when ancestor lang attribute changes

twilco opened the pull request:
#### 0b48826b8e58f154aae07a9895bf9f9290a8a926
<pre>
AX: AXPropertyName::Language can become out-of-date when ancestor lang attribute changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=270833">https://bugs.webkit.org/show_bug.cgi?id=270833</a>
<a href="https://rdar.apple.com/124424555">rdar://124424555</a>

Reviewed by Chris Fleizach.

Prior to this patch, we only updated AXPropertyName::Language when the lang attribute changed on that node. But
AccessibilityObject::language() iterates the ancestry, meaning this property is also dependent on its ancestors lang
attribute values (and the meta tag&apos;s content-language).

To fix this, AXPropertyName::Language now only represents the language of the individual object, and a new function is
added: AXCoreObject::languageIncludingAncestors(). This new function is used in all the places that expected the
language-of-this-or-ancestor functionality.

This also inherently improves performance, as we no longer do a full ancestry walk to compute AXPropertyName::Language
for every single node change, instead doing it on-demand to serve AT requests on the secondary thread.

Finally, this patch improves iOS test coverage by enabling language-attribute-change.html on this platform, and
implementing AccessibilityUIElement::language() so it can pass.

* LayoutTests/accessibility/language-attribute-change-expected.txt:
* LayoutTests/accessibility/language-attribute-change.html:
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::languageIncludingAncestors const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::language const):
* Source/WebCore/accessibility/AccessibilitySVGObject.cpp:
(WebCore::AccessibilitySVGObject::childElementWithMatchingLanguage const):
* Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::locale const):
* Source/WebCore/accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::textAttributes const):
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::attributeStringSetLanguage):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityLanguage]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
(WTR::AccessibilityUIElement::language):

Canonical link: <a href="https://commits.webkit.org/292610@main">https://commits.webkit.org/292610@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2366a18a66e65407569b5530bd38d42ccfd859c1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96257 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15871 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5840 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101324 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46776 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98302 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16167 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24304 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73372 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30603 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99260 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12130 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86953 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53711 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11881 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4714 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46103 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81999 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4812 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103352 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23324 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16981 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82411 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23575 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82972 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81787 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20592 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26417 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3854 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16698 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23287 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28442 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22946 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26426 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24687 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -122,6 +122,8 @@ void AXIsolatedObject::initializeProperties(const Ref<AccessibilityObject>& axOb
             setProperty(AXProperty::ListMarkerLineID, object.listMarkerLineID());
         }
+
+        setProperty(AXProperty::Language, object.language().isolatedCopy());
```
does this need to be gated by text thread apis?

If language() isEmpty() do we want to avoid doing an isolated copy and store?

twilco replied:
> does this need to be gated by text thread apis?

Ah, no it doesn't, good catch!

> If language() isEmpty() do we want to avoid doing an isolated copy and store?

Fixed!

hoffmanjoshua commented on LayoutTests/accessibility/language-attribute-change.html:
```diff
@@ -1,34 +1,44 @@
+    var accessibleContent = accessibilityController.accessibleElementById("content");
+    output += expect("accessibleContent.language", "'AXLanguage: en'");
+
+    content.innerText = "Hola Mundo!";
```
I know you are just modifying this existing test, but do you know why it is necessary to update the inner text when the `lang` attribute changes? If it's for decoration, I wonder if adding more property changes could mask over any issues updating `Language` in the cache?

twilco replied:
Good point, yeah there's no real reason to do this. I'll remove it from the test.

hoffmanjoshua replied:
Thanks! Otherwise looks great

twilco committed:
AX: AXPropertyName::Language can become out-of-date when ancestor lang attribute changes
https://bugs.webkit.org/show_bug.cgi?id=270833
rdar://124424555

Reviewed by Chris Fleizach.

Prior to this patch, we only updated AXPropertyName::Language when the lang attribute changed on that node. But
AccessibilityObject::language() iterates the ancestry, meaning this property is also dependent on its ancestors lang
attribute values (and the meta tag's content-language).

To fix this, AXPropertyName::Language now only represents the language of the individual object, and a new function is
added: AXCoreObject::languageIncludingAncestors(). This new function is used in all the places that expected the
language-of-this-or-ancestor functionality.

This also inherently improves performance, as we no longer do a full ancestry walk to compute AXPropertyName::Language
for every single node change, instead doing it on-demand to serve AT requests on the secondary thread.

Finally, this patch improves iOS test coverage by enabling language-attribute-change.html on this platform, and
implementing AccessibilityUIElement::language() so it can pass.

* LayoutTests/accessibility/language-attribute-change-expected.txt:
* LayoutTests/accessibility/language-attribute-change.html:
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::languageIncludingAncestors const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::language const):
* Source/WebCore/accessibility/AccessibilitySVGObject.cpp:
(WebCore::AccessibilitySVGObject::childElementWithMatchingLanguage const):
* Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::locale const):
* Source/WebCore/accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::textAttributes const):
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::attributeStringSetLanguage):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityLanguage]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
(WTR::AccessibilityUIElement::language):

Canonical link: https://commits.webkit.org/292610@main
